docling/datamodel/pipeline_options.py

```diff
@@ -1,8 +1,62 @@
+import logging
+import os
 from enum import Enum
 from pathlib import Path
-from typing import List, Literal, Optional, Union
+from typing import Any, Dict, List, Literal, Optional, Tuple, Type, Union
 
 from pydantic import BaseModel, ConfigDict, Field
+from pydantic_settings import (
+    BaseSettings,
+    PydanticBaseSettingsSource,
+    SettingsConfigDict,
+)
+
+_log = logging.getLogger(__name__)
+
+
+class AcceleratorDevice(str, Enum):
+    """Devices to run model inference"""
+
+    AUTO = "auto"
+    CPU = "cpu"
+    CUDA = "cuda"
+    MPS = "mps"
+
+
+class AcceleratorOptions(BaseSettings):
+    model_config = SettingsConfigDict(
+        env_prefix="DOCLING_", env_nested_delimiter="_", populate_by_name=True
+    )
+
+    # num_threads: int = Field(default=4, validation_alias="omp_num_threads")
+    num_threads: int = 4
+    device: AcceleratorDevice = AcceleratorDevice.AUTO
+
+    def __init__(self, **data):
+        r"""
+        Set num_threads from the "alternative" envvar OMP_NUM_THREADS.
+        The alternative envvar is used only if it is valid and the regular envvar is not set.
+
+        Notice: The standard pydantic settings mechanism with parameter "aliases" does not provide
+        the same functionality. In case the alias envvar is set and the user tries to override the
+        parameter in settings initialization, Pydantic treats the parameter provided in __init__()
+        as an extra input instead of simply overwriting the evvar value for that parameter.
+        """
+        input_num_threads = data.get("num_threads")
+
+        # Check if to set the num_threads from the alternative envvar
+        if input_num_threads is None:
+            docling_num_threads = os.getenv("DOCLING_NUM_THREADS")
+            omp_num_threads = os.getenv("OMP_NUM_THREADS")
+            if docling_num_threads is None and omp_num_threads is not None:
+                try:
+                    data["num_threads"] = int(omp_num_threads)
+                except ValueError:
+                    _log.error(
+                        "Ignoring misformatted envvar OMP_NUM_THREADS '%s'",
+                        omp_num_threads,
+                    )
```

Comment:
This should be done in the field validator not the `__init__`, which should already have the standard ENV resolved.

Reply:
Please check [my comments](https://github.com/DS4SD/docling/blob/6f0b91287cb43d0b4b13c61d8f4d8ab51dc04894/docling/datamodel/pipeline_options.py#L40). Pydantic raises an exception for an "extra" field if you let it do the standard resolution (under the conditions I describe in my comments). Therefore we cannot use a field validator and any custom code must run before the resolution and have access to the full model input.
As an alternative approach I can implement the same logic inside a model validator.

Reply:
Assuming this is really needed, it cannot be in the `__init__`, because now this is breaking all the argument resolution and validation (see example IDE).

Let's try the model validator.

![image](https://github.com/user-attachments/assets/8b402cd4-1786-4e46-911f-f1734cbbf482)